scripts/west_commands/patch.py

```diff
@@ -0,0 +1,327 @@
+            """
+            ),
+        )
+
+        subparsers.add_parser(
+            "list",
+            help="List patches",
+            formatter_class=argparse.RawDescriptionHelpFormatter,
+            epilog=textwrap.dedent(
+                """
+            Listing Patches:
+
+                Run "west patch list" to list patches.
+            """
+            ),
+        )
+
+        return parser
+
+    def filter_args(self, args):
+        manifest_path = None
+
+        try:
+            manifest_path = self.config.get("manifest.path")
+        except BaseException:
+            self.die("could not retrieve manifest path from west configuration")
+
+        # pathlib objects are not iterable (str in obj does not work)
+        patch_base = str(args.patch_base)
+        patch_yml = str(args.patch_yml)
+        west_workpace = str(args.west_workspace)
+        manifest_dir = str(Path(self.topdir) / Path(manifest_path))
+
+        if "WEST_MANIFEST_DIR" in patch_base:
+            patch_base = patch_base.replace("WEST_MANIFEST_DIR", manifest_dir)
+            args.patch_base = Path(patch_base)
+        if "WEST_MANIFEST_DIR" in patch_yml:
+            patch_yml = patch_yml.replace("WEST_MANIFEST_DIR", manifest_dir)
+            args.patch_yml = Path(patch_yml)
+        if "WEST_TOPDIR" in west_workpace:
+            west_workpace = west_workpace.replace("WEST_TOPDIR", self.topdir)
+            args.west_workspace = Path(west_workpace)
+
+    def do_run(self, args, _):
+        self.filter_args(args)
+
+        if not os.path.isfile(args.patch_yml):
+            self.inf(f"no patches to apply: {args.patch_yml} not found")
+            return
+
+        west_config = Path(args.west_workspace) / ".west" / "config"
+        if not os.path.isfile(west_config):
+            self.die(f"{args.west_workspace} is not a valid west workspace")
+
+        try:
+            with open(args.patch_yml) as f:
+                yml = yaml.load(f, Loader=SafeLoader)
+            pykwalify.core.Core(source_data=yml, schema_data=patches_schema).validate()
+        except (yaml.YAMLError, pykwalify.errors.SchemaError) as e:
+            self.die(f"ERROR: Malformed yaml {args.patch_yml}: {e}")
+
+        if not args.subcommand:
+            args.subcommand = "list"
+
+        method = {
+            "apply": self.apply,
+            "clean": self.clean,
+            "list": self.list,
+        }
+
+        method[args.subcommand](args, yml)
+
+    def apply(self, args, yml):
+        if not (yml and "patches" in yml and yml["patches"]):
+            self.inf("no patches to apply")
+            return
+
+        patch_count = 0
+        failed_patch = None
+        patched_mods = set()
+
+        for patch_info in yml["patches"]:
+            pth = patch_info["path"]
+            patch_path = os.path.realpath(Path(args.patch_base) / pth)
+
+            apply_cmd = patch_info["apply-command"]
+            apply_cmd_list = apply_cmd.split()
```

Comment:
Might be "safer" to use `shlex.split` here? In the case of some odd command with quoted arguments.

Reply:
Fixed!